### PR TITLE
Feature: Add `--probSpecsDir` option

### PR DESCRIPTION
**Pull request by ee7**
See the commit messages for more information - the commits are atomic.

Some things like the `git` error messages could probably be cleaned up a bit, and there's some obvious tests to add, but it should be reasonably robust. I think the implementation here is quite strict. Let me know if I missed any edge cases.

WIP:
- [X] Don't `checkout` or `merge`, even if the working directory is clean.
- [X] Try not to assume things about the remotes in the given directory.

Closes: https://github.com/exercism/canonical-data-syncer/issues/47

### CODE REVIEW

File: tests/test_probspecs.nim [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# This module contains tests for `src/probspecs.nim`
+import std/[json, options, os, osproc, strformat, unittest]
+import cli, probspecs
+
+type
+  ProblemSpecsDir = enum
+    psFresh = "fresh clone"
+    psExisting = "existing dir (simulate the `--probSpecsDir` option)"
+
+proc main =
```
Comment (ErikSchierboom) [inline tests/test_probspecs.nim:10]: So good to finally have some tests!

File: src/probspecs.nim [hunk 2]
```diff
@@ -100,13 +102,70 @@ proc findProbSpecsExercises(repo: ProbSpecsRepo, conf: Conf): seq[ProbSpecsExerc
...
+      # upstream, even if it's possible to do a fast-forward merge.
+      if execCmd(&"git fetch --quiet {remoteName} {mainBranchName}") != 0:
+        showError(&"failed to fetch `{mainBranchName}` in problem-specifications directory: '{probSpecsDir}'")
+
+      # Allow HEAD being on a non-`master` branch, as long as it's up-to-date
+      # with `upstream/master`.
+      let (revHead, _) = execCmdEx("git rev-parse HEAD")
+      let (revUpstream, _) = execCmdEx(&"git rev-parse {remoteName}/{mainBranchName}")
+      if revHead != revUpstream:
+        showError(&"the given problem-specifications directory is not up-to-date: '{probSpecsDir}'")
```
Comment (ErikSchierboom) [inline src/probspecs.nim:161]: Nice! I would suggest moving this to a separate function, as the verification code now slightly obscures the flow of the function

File: src/probspecs.nim [hunk 3]
```diff
@@ -100,13 +102,70 @@ proc findProbSpecsExercises(repo: ProbSpecsRepo, conf: Conf): seq[ProbSpecsExerc
...
+
+    withDir probSpecsDir:
+      if execCmd("git rev-parse") != 0:
+        showError(&"the given problem-specifications directory is not a git repository: '{probSpecsDir}'")
+
+      # Exit if the working directory is not clean.
+      if execCmd("git diff-index --quiet HEAD") != 0: # Ignores untracked files.
+        echo &"\nUnstaged changes in {probSpecsDir}:"
+        discard execCmd("git status --short --untracked=no")
+        showError(&"the given problem-specifications working directory is not clean: '{probSpecsDir}'")
```
Comment (ErikSchierboom) [inline src/probspecs.nim:132]: Maybe output this the other way around?

```suggestion
        showError(&"the given problem-specifications working directory has unstaged changes: '{probSpecsDir}'")
        discard execCmd("git status --short --untracked=no")
```

**Review by ErikSchierboom — APPROVED** [review]
Some minor comments! Excellent work.

File: src/probspecs.nim (hunk 3 above)
Comment (ee7) [reply src/probspecs.nim:132]: Doesn't work, unfortunately. `showError` isn't the best name, but it causes an exit.

But I agree that it might be better to see the `git status` output after the error, rather than before.

Comment (ErikSchierboom) [reply src/probspecs.nim:132]: An alternative would be to not display the unstaged changes.

File: tests/test_probspecs.nim (hunk 1 above)
Comment (ee7) [reply tests/test_probspecs.nim:10]: Agreed.

File: src/probspecs.nim (hunk 2 above)
Comment (ee7) [reply src/probspecs.nim]: Good idea. Done.

I also moved the remote-finding into its own function, but didn't do that for the rest of the validating. It think it might be better to put it into a new module if we wanted to do that.

File: src/probspecs.nim (hunk 3 above)
Comment (ee7) [reply src/probspecs.nim]: Done.

File: src/probspecs.nim [hunk 4]
```diff
@@ -100,13 +102,87 @@ proc findProbSpecsExercises(repo: ProbSpecsRepo, conf: Conf): seq[ProbSpecsExerc
...
+    # Exit if the given directory is not a git repo.
+    if execCmd("git rev-parse") != 0:
+      showError("the given problem-specifications directory is not a git " &
+                &"repository: '{probSpecsDir}'")
+
+    # Exit if the working directory is not clean.
+    if execCmd("git diff-index --quiet HEAD") != 0: # Ignores untracked files.
+      echo &"\nUnstaged changes in {probSpecsDir}:"
+      showError("the given problem-specifications working directory is not " &
+                &"clean: '{probSpecsDir}'")
```
Comment (ErikSchierboom) [inline src/probspecs.nim:154]:
```suggestion
      showError("the given problem-specifications working directory has" &
                &"unstaged changes: '{probSpecsDir}'")
```

File: src/probspecs.nim [hunk 5]
```diff
@@ -100,13 +102,87 @@ proc findProbSpecsExercises(repo: ProbSpecsRepo, conf: Conf): seq[ProbSpecsExerc
...
+    let (revHead, _) = execCmdEx("git rev-parse HEAD")
+    let (revUpstream, _) = execCmdEx(&"git rev-parse {remoteName}/{mainBranchName}")
+    if revHead != revUpstream:
+      showError("the given problem-specifications directory is not " &
+                &"up-to-date: '{probSpecsDir}'")
+
+proc findProbSpecsExercises*(conf: Conf): seq[ProbSpecsExercise] =
+  if conf.probSpecsDir.isSome():
+    let probSpecsRepo = ProbSpecsRepo(dir: conf.probSpecsDir.get())
+    validateProbSpecsRepo(probSpecsRepo)
```
Comment (ErikSchierboom) [inline src/probspecs.nim:179]: The other calls in this function use `probSpecsRepo.<function>()`. Was there any particular reason not to do that here?

```suggestion
    probSpecsRepo.validateProbSpecsRepo()
```

Mayb even do:

```suggestion
    probSpecsRepo.validate()
```

**Review by ErikSchierboom — APPROVED** [review]

File: src/probspecs.nim (hunk 5 above)
Comment (ee7) [reply src/probspecs.nim]: Done, thanks.

File: src/probspecs.nim (hunk 4 above)
Comment (ee7) [reply src/probspecs.nim]: Thanks. Fixed.
